python/fastfec.py

```diff
@@ -1,59 +1,386 @@
+    def convert_item(i):
+        item = array_get(items, i, None)
+        if item is None:
+            return None
+        if types is None:
+            return item
+        fec_type = array_get(types, i, ord(b's'))
+        if fec_type == ord(b's'):
+            return item
+        if fec_type == ord(b'd'):
+            # Convert standard YYYY-MM-DD date to Pythonic date object
+            return datetime.date(*(int(component) for component in item.split('-')))
+        if fec_type == ord(b'f'):
+            return float(item)
+
+        logging.warning("Unrecognized type: %s", chr(fec_type))
+        return item
+
+    # Build up result object
+    result = {}
+    # Used to handle missing header #'s
+    # (we don't expect this to happen, but FEC filings sometimes have
+    # more values than headers in a particular row. If this happens,
+    # we can still capture the data with a `__missing_header_#` key)
+    missing_header = 1
+    for i in range(max(len(headers), len(items))):
+        if i < len(headers):
+            header = headers[i]
+        else:
+            header = f'__missing_header_{missing_header}'
+            missing_header += 1
+        item = convert_item(i)
+        result[header] = item
+
+    return result
+
+
+def provide_read_callback(file_handle):
+    """Provides a C callback to read from a given file stream"""
+    return BUFFER_READ(
+        make_read_buffer(file_handle))
+
+
+def provide_write_callback(open_function):
+    """Provides a C callback to write to file given a function to open file streams"""
+    # Initialize parsing cache
+    write_cache = WriteCache()
+
+    def write_callback(filename, extension, contents, num_bytes):
+        if filename == write_cache.last_filename:
+            # Same filename as last call? Reuse file handle
+            write_cache.last_fd.write(contents[:num_bytes])
+        else:
+            path = filename + extension
+            # Grab the file descriptor from the cache if possible
+            file_descriptor = write_cache.file_descriptors.get(path)
+            if not file_descriptor:
+                # Open the file
+                write_cache.file_descriptors[path] = open_function(
+                    path.decode("utf8"), mode='wb')
+                file_descriptor = write_cache.file_descriptors[path]
+            write_cache.last_filename = filename
+            write_cache.last_fd = file_descriptor
+            # Write to the file descriptor
+            file_descriptor.write(contents[:num_bytes])
+
+    def free_file_descriptors():
+        for path in write_cache.file_descriptors:
+            write_cache.file_descriptors[path].close()
+
+    return [CUSTOM_WRITE(write_callback), free_file_descriptors]
+
+
+def provide_line_callback(queue):
+    """Provides a C callback to return parsed lines given a queue to handle threading
+
+    The threading allows the parent caller to yield result lines as they are returned"""
```

Comment:
I have a lot of questions about the how and why of this, it might be easier to zoom?